core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -332,6 +332,12 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
 
             switch (node.getType()) {
                 case GATHER:
+                    if (inputProperties.size() == 1) {
+                        StreamProperties properties = inputProperties.get(0);
+                        if (properties.isSingleStream() && properties.isOrdered()) {
```

Comment:
@kasiafi is `isSingleStream()` equivalent to `isEffectivelySingleStream`. Do we need to have both methods or maybe `isSingleStream` could delegate to `isEffectivelySingleStream`.

I wonder if there is a case where we loose local sorting properties because `isEffectivelySingleStream=false`, but `isSingleStream` is still true.

Reply:
I think it could be the other way. It depends on whether the `effectivelySingleStream` plan (i.e. partitioned on constants) is actually planned without  repartition.

Reply:
`isEffectivelySingleStream` checks `streamPartitioning`. Can there be unpartitioned single stream? Then `isEffectivelySingleStream` would be `false`, but `isSingleStream` would be true.